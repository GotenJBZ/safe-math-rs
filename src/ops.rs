use crate::error::SafeMathError;
use std::ops::{Add, Div, Mul, Rem, Sub};
#[diagnostic::on_unimplemented(
    message = "Type `{Self}` cannot perform safe addition.",
    note = "Add `add` to `#[SafeMathOps(...)]` when deriving `SafeMathOps`."
)]
pub trait SafeAdd: Copy + Add<Output = Self> {
    fn safe_add(self, rhs: Self) -> Result<Self, SafeMathError>;
}
#[diagnostic::on_unimplemented(
    message = "Type `{Self}` cannot perform safe subtraction.",
    note = "Add `sub` to `#[SafeMathOps(...)]` when deriving `SafeMathOps`."
)]
pub trait SafeSub: Copy + Sub<Output = Self> {
    fn safe_sub(self, rhs: Self) -> Result<Self, SafeMathError>;
}

#[diagnostic::on_unimplemented(
    message = "Type `{Self}` cannot perform safe multiplication.",
    note = "Add `mul` to `#[SafeMathOps(...)]` when deriving `SafeMathOps`."
)]
pub trait SafeMul: Copy + Mul<Output = Self> {
    fn safe_mul(self, rhs: Self) -> Result<Self, SafeMathError>;
}

#[diagnostic::on_unimplemented(
    message = "Type `{Self}` cannot perform safe division.",
    note = "Add `div` to `#[SafeMathOps(...)]` when deriving `SafeMathOps`."
)]
pub trait SafeDiv: Copy + Div<Output = Self> {
    fn safe_div(self, rhs: Self) -> Result<Self, SafeMathError>;
}

#[diagnostic::on_unimplemented(
    message = "Type `{Self}` cannot perform safe remainder operation.",
    note = "Add `rem` to `#[SafeMathOps(...)]` when deriving `SafeMathOps`."
)]
pub trait SafeRem: Copy + Rem<Output = Self> {
    fn safe_rem(self, rhs: Self) -> Result<Self, SafeMathError>;
}

pub trait SafeMathOps: Copy {
    fn safe_add(self, rhs: Self) -> Result<Self, SafeMathError>;
    fn safe_sub(self, rhs: Self) -> Result<Self, SafeMathError>;
    fn safe_mul(self, rhs: Self) -> Result<Self, SafeMathError>;
    fn safe_div(self, rhs: Self) -> Result<Self, SafeMathError>;
    fn safe_rem(self, rhs: Self) -> Result<Self, SafeMathError>;
}
