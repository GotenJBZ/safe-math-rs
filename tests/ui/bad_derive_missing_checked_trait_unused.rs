use safe_math::SafeMathOps;
use std::ops::Add;

#[derive(Debug,Copy,Clone,SafeMathOps)]
#[SafeMathOps(add)]
struct Foo(i32);

impl Add for Foo {
    type Output = Foo;
    fn add(self, other: Foo) -> Foo {
        Foo(self.0 + other.0)
    }
}

fn main() {} 