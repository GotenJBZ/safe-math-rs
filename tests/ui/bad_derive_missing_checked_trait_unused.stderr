error[E0277]: the trait bound `Foo: CheckedAdd` is not satisfied
 --> tests/ui/bad_derive_missing_checked_trait_unused.rs:4:27
  |
4 | #[derive(Debug,Copy,Clone,SafeMathOps)]
  |                           ^^^^^^^^^^^ the trait `CheckedAdd` is not implemented for `Foo`
  |
  = help: the following other types implement trait `CheckedAdd`:
            i128
            i16
            i32
            i64
            i8
            isize
            u128
            u16
          and $N others
  = help: see issue #48214
  = note: this error originates in the derive macro `SafeMathOps` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: Type `Foo` cannot perform safe addition.
 --> tests/ui/bad_derive_missing_checked_trait_unused.rs:4:27
  |
4 | #[derive(Debug,Copy,Clone,SafeMathOps)]
  |                           ^^^^^^^^^^^ the trait `SafeAdd` is not implemented for `Foo`
  |
  = note: Add `add` to `#[SafeMathOps(...)]` when deriving `SafeMathOps`.
  = note: this error originates in the derive macro `SafeMathOps` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no method named `checked_add` found for struct `Foo` in the current scope
 --> tests/ui/bad_derive_missing_checked_trait_unused.rs:4:27
  |
4 | #[derive(Debug,Copy,Clone,SafeMathOps)]
  |                           ^^^^^^^^^^^ method not found in `Foo`
5 | #[SafeMathOps(add)]
6 | struct Foo(i32);
  | ---------- method `checked_add` not found for this struct
  |
 ::: $CARGO/num-traits-0.2.19/src/ops/checked.rs
  |
  |     fn checked_add(&self, v: &Self) -> Option<Self>;
  |        ----------- the method is available for `Foo` here
  |
  = help: items from traits can only be used if the trait is in scope
  = note: this error originates in the derive macro `SafeMathOps` (in Nightly builds, run with -Z macro-backtrace for more info)
help: trait `CheckedAdd` which provides `checked_add` is implemented but not in scope; perhaps you want to import it
  |
1 + use num_traits::CheckedAdd;
  |
